toolkit/tools/pkg/srpmpacker/srpmpacker.go

```diff
@@ -0,0 +1,994 @@
+package srpmpacker
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/buildpipeline"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/directory"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/file"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/jsonutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/network"
+	packagelist "github.com/microsoft/CBL-Mariner/toolkit/tools/internal/packlist"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/retry"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/rpm"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/timestamp"
+)
+
+const (
+	SignatureEnforceString   = "enforce"
+	SignatureSkipCheckString = "skip"
+	SignatureUpdateString    = "update"
+)
+
+type fileSignaturesWrapper struct {
+	FileSignatures map[string]string `json:"Signatures"`
+}
+
+const (
+	srpmOutDir     = "SRPMS"
+	srpmSPECDir    = "SPECS"
+	srpmSOURCESDir = "SOURCES"
+)
+
+type fileType int
+
+const (
+	fileTypePatch  fileType = iota
+	fileTypeSource fileType = iota
+)
+
+type signatureHandlingType int
+
+const (
+	signatureEnforce   signatureHandlingType = iota
+	signatureSkipCheck signatureHandlingType = iota
+	signatureUpdate    signatureHandlingType = iota
+)
+
+// sourceRetrievalConfiguration holds information on where to hydrate files from.
+type sourceRetrievalConfiguration struct {
+	localSourceDir string
+	sourceURL      string
+	caCerts        *x509.CertPool
+	tlsCerts       []tls.Certificate
+
+	signatureHandling signatureHandlingType
+	signatureLookup   map[string]string
+}
+
+// packResult holds the worker results from packing a SPEC file into an SRPM.
+type packResult struct {
+	specFile string
+	srpmFile string
+	err      error
+}
+
+// specState holds the state of a SPEC file: if it should be packed and the resulting SRPM if it is.
+type specState struct {
+	specFile string
+	srpmFile string
+	toPack   bool
+	err      error
+}
+
+func (cfg *Config) CreateAllSrpmsWrapper() error {
```

Comment:
I don't think this should be a member function of `Config`. It should take `Config` as its first parameter,

Reply:
This was an intentional design choice and ofcourse is up for discussion.
The rationale behind this is our public interface function should have a `context` & maybe a `logger Interface` and rest flows via the config.
The `context` gives us the capability to bail out prematurely if we want. Having a `logger Interface` gives us capability to switch loggers and also maybe have a no-op logger for testing.
As it stands right now we have a global logger which is very closely coupled with the `logrus` library.
Also a few local experiments that I have done, give an indication that our logger implementation has an explicit `Mutex`;
If we have an API design that allows us to switch logger implementations at fly, there is a chance we would get some more perf.
What do you think ?

Reply:
There isn’t a limit to the number of parameters a function can take.

For example:

```golang
func (ctx context.Context, logger logrus.StdLogger, cfg *Config) { … }
```

The question at hand is if this is an operation on a config object or an operation that uses the config object as input? I’m pretty sure it is the latter.

Reply:
There have been some cases where it would have been handy if we supported `context` in the toolkit. But fixing this will be a very large undertaking. And I’m not sure the benefit justifies the cost at the moment.

`context` really comes in handy when you are doing a lot of network operations, which may fail or you may want to cancel for taking too long. But a lot of what the toolkit does doesn’t use the network.

Reply:
In regards to logging, not needing to pass around a logger object is somewhat nice and it reduces visual noise within the codebase. Also, changing the logging pattern would be a massive undertaking. So, there would need to be strong justification for changing it.

If you think there is performance improvements to be had, then you could write an experimental change and provide the % improvement you observed.